libraries/apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/codegen/kotlin/helpers/DataClass.kt

```diff
@@ -62,17 +65,193 @@ fun TypeSpec.Builder.makeDataClassFromProperties(properties: List<PropertySpec>)
+        .beginControlFlow("if (%L == null)", MEMOIZED_HASH_CODE_VAR)
+        .addStatement("var ${Identifier.__h} = %L.hashCode()", propertySpecs.getOrNull(0)?.name ?: "null")
+        .add(propertySpecs
+            .drop(1)
+            .excludeInternalProperties()
+            .map(::hashPropertyCode)
+            .fold(CodeBlock.builder(), CodeBlock.Builder::add)
+            .build())
+        .addStatement("%L = ${Identifier.__h}", MEMOIZED_HASH_CODE_VAR)
+        .endControlFlow()
+        .addStatement("return %L!!", MEMOIZED_HASH_CODE_VAR)
+        .build()
+  }
+
+  check(propertySpecs.all { !it.mutable }) {
+    "Cannot generate hashCode function for class with mutable properties, memoization would fail"
+  }
+
+  if (propertySpecs.isNotEmpty()) {
+    addProperty(
+        PropertySpec.builder(MEMOIZED_HASH_CODE_VAR, KotlinSymbols.Int.copy(nullable = true), KModifier.PRIVATE)
+            .mutable()
+            .initializer("null")
+            .build()
+    )
+  }
+
+  addFunction(FunSpec.builder(Identifier.hashCode)
+      .addModifiers(KModifier.OVERRIDE)
+      .returns(KotlinSymbols.Int)
+      .addCode(methodCode())
+      .build()
+  )
+}
+
+fun TypeSpec.Builder.withToStringImplementation(): TypeSpec.Builder {
+  val name = build().name ?: return this
+  fun printPropertiesTemplate() =
+      "$name(" + propertySpecs
+            .excludeInternalProperties()
+            .joinToString(",") { "${it.name}=\$${it.name}" } + ")"
+
+  fun methodCode() =
+      CodeBlock.builder()
+          .beginControlFlow("if (%L == null)", MEMOIZED_TO_STRING_VAR)
+          .add("%L = %P", MEMOIZED_TO_STRING_VAR, printPropertiesTemplate())
+          .endControlFlow()
+          .addStatement("return %L!!", MEMOIZED_TO_STRING_VAR)
+          .build()
+
+  return addProperty(
+      PropertySpec.builder(MEMOIZED_TO_STRING_VAR, KotlinSymbols.String.copy(nullable = true), KModifier.PRIVATE)
+          .mutable()
+          .initializer("null")
+          .build()
+  )
+      .addFunction(FunSpec.builder(Identifier.toString)
+          .addModifiers(KModifier.OVERRIDE)
+          .returns(KotlinSymbols.String)
+          .addCode(methodCode())
+          .build())
+}
+
+
+private fun Collection<PropertySpec>.excludeInternalProperties(): Collection<PropertySpec> {
+  return filterNot { it.name == MEMOIZED_HASH_CODE_VAR || it.name == MEMOIZED_TO_STRING_VAR }
 }
 
-private fun equalsFunSpec() = FunSpec.builder(Identifier.equals)
-    .addModifiers(KModifier.OVERRIDE)
-    .addParameter("other", KotlinSymbols.Any.copy(nullable = true))
-    .returns(KotlinSymbols.Boolean)
-    .addStatement("return·other != null·&&·other::class·==·this::class")
-    .build()
-
-private fun hashCodeFunSpec() = FunSpec.builder(Identifier.hashCode)
-    .addModifiers(KModifier.OVERRIDE)
-    .returns(KotlinSymbols.Int)
-    .addStatement("return·this::class.hashCode()")
-    .build()
+private const val MEMOIZED_HASH_CODE_VAR: String = "__hashCode"
```

Comment:
I added memoization because the Java implementation had it. I added a check that the properties are immutable for safety.